src/saas_shield/deterministic.rs

```diff
@@ -175,6 +161,83 @@ impl DeterministicFieldOps for SaasShieldDeterministicClient {
             .collect()
     }
 
+    async fn rotate_fields(
+        &self,
+        encrypted_fields: EncryptedFields,
+        metadata: &AlloyMetadata,
+        new_tenant_id: TenantId,
+    ) -> Result<RotateBatchResult, AlloyError> {
+        let paths = encrypted_fields
+            .values()
+            .map(|field| (field.secret_path.clone(), field.derivation_path.clone()))
+            .collect_vec();
+        let original_tenant_keys = derive_keys_many_paths(
+            &self.tenant_security_client,
+            metadata,
+            paths.clone(),
+            SecretType::Deterministic,
+        )
+        .await?;
+        let new_tenant_keys = if new_tenant_id != metadata.tenant_id {
+            derive_keys_many_paths(
+                &self.tenant_security_client,
+                metadata,
+                paths,
+                SecretType::Deterministic,
+            )
+            .await?
+        } else {
+            original_tenant_keys.clone() // TODO: I only need a reference to this, but the `else` creates an owned value.
+                                         // Is there a better way to do it?
+        };
+        let reencrypt_field = |encrypted_field: EncryptedField| {
+            let (original_key_id, ciphertext) =
+                Self::decompose_encrypted_field_header(encrypted_field.encrypted_field.clone())?;
+            let original_key = derive_key_for_path(
+                &original_tenant_keys,
+                &encrypted_field.secret_path,
+                &encrypted_field.derivation_path,
+                DeriveKeyChoice::Specific(original_key_id),
+            )?;
+            let new_current_key = derive_key_for_path(
+                &new_tenant_keys,
+                &encrypted_field.secret_path,
+                &encrypted_field.derivation_path,
+                DeriveKeyChoice::Current,
+            )?;
+            if check_rotation_no_op(
+                original_key_id,
+                &Some(new_current_key.tenant_secret_id.0),
+                &new_tenant_id,
+                metadata,
+            ) {
+                Ok(encrypted_field)
+            } else {
+                let decrypted_field = decrypt_internal(
+                    DeterministicEncryptionKey(original_key.derived_key.0.clone()),
+                    ciphertext,
+                    encrypted_field.secret_path.clone(),
+                    encrypted_field.derivation_path.clone(),
+                )?;
+                let key_id_header = KeyIdHeader::new(
```

Comment:
Should we add a function to the AlloyClient which will make a KeyIdHeader given a KeyId?

Reply:
Good idea. Done